### ISSUE TITLE
Area2Ds may not overlap

### ISSUE THREAD

**Issue by mjtorn** [godotengine/escoria#221]
Please see #213 for details. Input events get sprayed just about wherever, causing mad problems if Area2Ds overlap.

------

I solved this by replacing an underlying Area2D - which fortunately was passive anyway - with a Node2D, as equally fortunately the important stuff is in an animation node anyway.

**Comment by fleskesvor** [comment]
I'm labelling this as a bug, since it might be considered a regression stemming all the way back from #8. There are considerable advantages to using `Area2D`s for the background and items, but we haven't yet resolved the issues caused by using nodes which don't block input.

**Comment by mjtorn** [comment]
This is buggy as anything. I noticed during recent development, that a `mouse_exited` can come after a `mouse_entered` and there's no apparent way of enforcing an order for events.

Thus I added some accounting for potentially overlapped objects. This serves the purpose of having items overlapped by collapsible inventories and such.

I can't tell if that caused a regression in overlapping areas or if they have still always been broken - near-burnout stress destroyed my memory and there's so much back-and-forth between areas, tooltips, dialog and everything to make it all work... - but currently the problem is that entering an overlapping area does not cause a `mouse_exited` to be emitted from the previous area.

The diff below alleviates the problem a bit by forcing a `mouse_exited` to be emitted, but it has the drawback that it doesn't recognize re-entering the previous object.

Because the mouse exits by emitting the signal, `overlaid_obj` will be set to `null` and that's why there's an extra check. This might be really redundant, I don't think the `if` will ever be true, but this is an experiment.

It also seemed to cause breakage in the first room of A Lunar Adventure, moving the mouse real quick between the control panel and the tool storage. Sometimes the tooltip would be missing, which is exactly the type of problem I was originally intending to fix.

It might be overkill, but would we need a push/pop stack of overlaid objects?

```diff
diff --git a/device/globals/game.gd b/device/globals/game.gd
index 06ec5b9..7c78a87 100644
--- a/device/globals/game.gd
+++ b/device/globals/game.gd
@@ -41,6 +41,7 @@ func set_overlapped_obj(obj):
 
 func reset_overlapped_obj():
 	if overlapped_obj:
+		printt("reset overlapped: ", overlapped_obj.name)
 		return mouse_enter(overlapped_obj)
 
 func maybe_hide_tooltip():
@@ -79,18 +80,27 @@ func tooltip_clamped_position(tt_pos):
 	return tt_pos
 
 func mouse_enter(obj):
+	printt("mouse_entered", obj, obj.name)
+	printt("\t", obj.get_tooltip())
 	if overlapped_obj and overlapped_obj != obj:
 		# Be sure we have exited the other object, because
 		# sometimes item2's `mouse_entered` happens before
 		# item1's `mouse_exited`. This causes the tooltip to disappear!
 		if overlapped_obj.has_node("area"):
-			yield(overlapped_obj.get_node("area"), "mouse_exited")
+			printt("mouse_enter yield: ", overlapped_obj.name, " area")
+			overlapped_obj.get_node("area").emit_signal("mouse_exited")
+			if overlapped_obj:
+				yield(overlapped_obj.get_node("area"), "mouse_exited")
 		else:
-			yield(overlapped_obj, "mouse_exited")
+			printt("mouse_enter yield: ", overlapped_obj.name)
+			overlapped_obj.emit_signal("mouse_exited")
+			if overlapped_obj:
+				yield(overlapped_obj, "mouse_exited")
 
 	# Immediately bail out if the action menu is open
 	if action_menu and action_menu.is_visible():
 		assert(!tooltip.visible)
+		printt("return for action menu")
 		return
 
 	# Store overlapped_obj just in case we try to open the inventory
@@ -177,6 +187,7 @@ func mouse_exit(obj):
 		vm.hover_end()
 
 	overlapped_obj = null
+	printt("mouse_exited", obj, obj.name)
 
 func clear_action():
 	current_tool = null
```

**Comment by mjtorn** [comment]
I would like to hear some comments on how to deal with crazily overlapping Areas, because I'm not sure I can think this through alone quickly, and so it'll drag on forever.

Like... `area_1` contains `area_2`, which contains `area_3`, that makes it possible to exit all areas (onto the background). Then `mouse_exited` happens only on `area_1` or `area_3`

Is it safe to manually "unwind" the stack by emitting `mouse_exited` for all of them when exiting through `area_3`?

Would it have to somehow block the code in `area_3` so that we emit the signals starting from `area_1`?

How do people usually solve this problem? Are other engines different (or better) with this? Because surely Escoria can't be the first framework in the world to encounter this problem.

**Comment by mjtorn** [comment]
I think I have a clue on this now, and all current solutions are wrong.

All sorts of overlap tracking and such are wrong, because I've come to realize Godot is real nice and all, but it doesn't give you much by design, like the double-clicking thing.

The mouse does literally enter all `Area2D` nodes and we have to live with that. I talked a bit about designing this with a friend, and if Z-indexes don't work or are a poor idea, the solution is to introduce a variable like `(int) interact_layer`.

Each `Area2D` will inhabit a layer and no two areas are allowed to overlap on the same layer.

This could potentially help us get rid of using the hack that looks for overlapping objects in the autogenerated background and maybe even using `TextureRect` in inventory.

Thoughts?

(This is not a priority at the moment, but closing this issue would be a huge win)

**Comment by StraToN** [comment]
> Each `Area2D`will inhabit a layer and no two areas are allowed to overlap on the same layer.

I don't understand this sentence, please rephrase?

IIUC, the problem is that the mouse input is passed through all overlapping `Area2D`?
This sounds like a [Mouse Filter subject](http://docs.godotengine.org/en/3.0/classes/class_control.html#enum-control-mousefilter). 

If a `Control` node (like `Area2D`) has `mouse_filter == MOUSE_FILTER_STOP`, then the input will not propagate to its **parents**. In the case where the 2 `Area2D` nodes are siblings and overlapping, this will not work.

So, I suppose that the hierarchy of `Area2D` nodes in the scene tree is important: the node that must stop the input to stop its propagation should be the deepest in the scene tree.

1 pro, 1 con:
- con: This solution forces the user to set the hierarchy of `Area2D` nodes in its own tree structure (having them in two different scenes doesn't matter if one is instanced as a (grand-*)child of the former. In other words: two `Area2D` nodes can't share the same parent.
- pro: Visually (ie. looking at the scene tree) it is easy to know which node is the input blocker: that's the deeper one.

This requires some testing. Also I just hope I understood the problem correctly.

**Comment by mjtorn** [comment]
What I was trying to propose is a design with "layers", which are basically just integers. When the mouse enters multiple areas, ie. overlapping areas, the code would see which one has the highest value and emit signals only on that.

That is the reason why there should be an `assert` or at least `vm.report_errors()` if two areas overlap with the same layer number.

With this design, the background can have a very low number, even `-1`, so any item (`>= 0`) will always get the priority.

Unfortunately you are mistaken about the mouse filter part. `Area2D` does _not_ inherit from `Control`, but `Node2D`. So basically anything related to that line of thinking (rearranging the tree when one NPC moves in front of another item or not having any overlaps) is moot.

**Comment by StraToN** [comment]
Ah indeed, this `Area2D` thing I described can't work. I wonder if there's another node that does the exact same thing but being a Control node.

Then I'm fine with idea you proposed, that should work.

**Comment by mjtorn** [comment]
Closing this, it's practically been closed by #258 for a while
